### ISSUE TITLE
Memory corruption in protobuf_c_message_free_unpacked

### ISSUE THREAD

**Issue by smuellerDD** [protobuf-c/protobuf-c#690]
The current git head  contains a memory corruption in protobuf_c_message_free_unpacked at the following line:

```
        if (message->unknown_fields != NULL)
                do_free(allocator, message->unknown_fields);
```

The do_free will free ->unknown_fields. This is usually wrong, because protobuf_c_message_unpack uses malloc as the default allocator, allocates `rv` with malloc. At the end, however, ->unknown_fields is only initialized if there are some:

```
        if (n_unknown) {
                rv->unknown_fields = do_alloc(allocator,
                                              n_unknown * sizeof(ProtobufCMessageUnknownField));
``` 

That means if there are no such fields ->unknown_fields is an uninitialized pointer.

As there are various ways to fix it, I did not send a patch yet. I see the following options:

- have an else branch during the allocation that sets ->unknown_fields to NULL
- use calloc as default allocator
- require the caller to use their own allocator to create a zeroized memory buffer (I guess that is not the right way)

I tested the first and last option with my code and it made the issue go away.

**Comment by edmonds** [comment]
Hi, @smuellerDD, thanks for the report and the patch.

I don't doubt that you ran into a memory corruption issue but I'm curious if you can provide more details about the exact circumstances that were required in order to get `protobuf_c_message_unpack()` to return a `ProtobufCMessage *` with an uninitialized `unknown_fields`? If protobuf-c were always returning uninitialized memory there I'd expect a lot more crashes and bug reports since that kind of bug would tend to light up Valgrind and the various sanitizers.

After the initial `do_alloc()` here:

https://github.com/protobuf-c/protobuf-c/blob/67b1c3535d909387e8319d6a085e7571a3ef8126/protobuf-c/protobuf-c.c#L3063

We perform initialization here:

https://github.com/protobuf-c/protobuf-c/blob/67b1c3535d909387e8319d6a085e7571a3ef8126/protobuf-c/protobuf-c.c#L3085-L3088

In the first branch, if the message descriptor was generated by the protobuf-c compiler we should eventually end up calling `PROTOBUF_C_MESSAGE_INIT` which will zero all three fields of the `ProtobufCMessage`:

https://github.com/protobuf-c/protobuf-c/blob/67b1c3535d909387e8319d6a085e7571a3ef8126/protobuf-c/protobuf-c.h#L999

In the second branch, we should be calling `memset()` to zero the whole thing:

https://github.com/protobuf-c/protobuf-c/blob/67b1c3535d909387e8319d6a085e7571a3ef8126/protobuf-c/protobuf-c.c#L2950

Are you by chance using your own `ProtobufCMessageDescriptor` rather than one generated by the protobuf-c compiler?

**Comment by edmonds** [comment]
Also, my preference for fixing this would be to immediately do a `*rv = {0};` right after the allocation succeeds, it's a little bit weird that the protobuf-c code base doesn't do this everywhere memory is allocated. Perhaps we should make `do_alloc()` always do a `memset` and take the probably negligible performance hit in exchange for more assurance that we're not using uninitialized memory somewhere.

**Comment by smuellerDD** [comment]
Am Sonntag, 21. Januar 2024, 16:39:46 CET schrieb Robert Edmonds:

Hi Robert,

> Also, my preference for fixing this would be to immediately do a `*rv =
> {0};` right after the allocation succeeds, it's a little bit weird that the
> protobuf-c code base doesn't do this everywhere memory is allocated.
> Perhaps we should make `do_alloc()` always do a `memset` and take the
> probably negligible performance hit in exchange for more assurance that
> we're not using uninitialized memory somewhere.

The easiest would be a calloc instead of a malloc.

Ciao
Stephan

**Comment by edmonds** [comment]
> The easiest would be a calloc instead of a malloc. Ciao Stephan

Right, but we can't directly call `calloc()` since protobuf-c wrap memory allocations using `ProtobufCAllocator` so that the library user can provide their own allocation functions. So we could either introduce a `do_calloc()` that calls the `alloc` function in the allocator and does a memset() or we could make `do_alloc()` always zero out all memory allocations.

**Comment by smuellerDD** [comment]
Am Sonntag, 21. Januar 2024, 16:31:56 CET schrieb Robert Edmonds:

Hi Robert,

> Are you by chance using your own `ProtobufCMessageDescriptor` rather than
> one generated by the protobuf-c compiler?

I am using a pretty standard code I would assume where seemingly I do not 
replace any data structure. However, I implement an RPC mechanism (one example 
of the implementation is in [1]). Yet, in addition, I wrap one protobuf 
definition into another to implement a cryptographic protocol around a regular 
RPC interface.

I.e. I have an outer protobuf call for the secure connection described in [3] 
followed by an internal protobuf definition with the various RPC calls I want 
to have.

The outer calls the inner with the ->invoke operation after the outer layer 
decrypted the data. In addition, when the inner layer completes, it writes the 
data into a buffer when calling protobuf_c_message_pack_to_buffer by providing 
my own base.append implementation.

In addition, for performance reason I have my own allocator which uses stack 
memory shown in [2].

Note, the funny thing is that I never experienced a crash, but valgrind showed 
me invalid memory accesses which I traced down to this location. Note, the 
invalid memory accesses were only off by up to 8 bytes. On Linux this usually 
does not cause a crash as this is somehow within the bounds of guard pages or 
similar mechanism that would prevent such crash.

[1] https://github.com/smuellerDD/esdm/blob/master/service-rpc/client/
esdm_rpc_client.c and https://github.com/smuellerDD/esdm/blob/master/service-rpc/server/esdm_rpc_server.c

[2] https://github.com/smuellerDD/esdm/blob/master/service-rpc/service/
esdm_rpc_protocol.c#L28

[3] http://www.chronox.de/papers/KyberKEX_2way_handshake_specification.pdf

Ciao
Stephan
